### PR TITLE
Qt: Copy General Tab from WX

### PR DESCRIPTION
**Pull request by Ryanel**
Adds the CPU Pane to settings: 

dolphinqt2d_2017-05-05_23-59-03

Notable differences:
- Dual-core is now called Multi-core. This is because Dolphin uses 3 cores (CPU, GPU, DSP), which is not 2.

### CODE REVIEW

**Comment by JosJuice** [comment]
Multi-core is not a more accurate name than dual core, as we use multiple cores even when the setting is turned off (DVD always has a separate thread, DSP sometimes has a separate thread, and I suppose there's things like controller input and memcard writing too...). I don't think it should be changed.

**Comment by Ryanel** [comment]
Changed to Dual-Core

**Comment by myownfriend** [comment]
Are the CPU Clock Override settings eventually going to be here, too?

File: Source/Core/DolphinQt2/Settings/CPUPane.cpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+void CPUPane::ConnectLayout()
+{
+  connect(m_checkbox_multicore, &QCheckBox::clicked, this, &CPUPane::OnSaveConfig);
+
+  connect(m_radio_interpreter, &QRadioButton::clicked, this, &CPUPane::OnSaveConfig);
+  connect(m_radio_cached_interpreter, &QRadioButton::clicked, this, &CPUPane::OnSaveConfig);
+  connect(m_radio_jit, &QRadioButton::clicked, this, &CPUPane::OnSaveConfig);
+  connect(m_radio_jitil, &QRadioButton::clicked, this, &CPUPane::OnSaveConfig);
```
Comment (goldenercobra) [inline Source/Core/DolphinQt2/Settings/CPUPane.cpp]: JITIL was removed, this is no longer necessary.

File: Source/Core/DolphinQt2/Settings/CPUPane.cpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  auto* engine_group = new QGroupBox(tr("Emulation Engine"));
+  auto* engine_group_layout = new QVBoxLayout;
+  engine_group->setLayout(engine_group_layout);
+  m_main_layout->addWidget(engine_group);
+
+  // CPU Emulator Engine
+  m_radio_interpreter = new QRadioButton(tr("Interpreter (slowest)"));
+  m_radio_cached_interpreter = new QRadioButton(tr("Cached Interpreter (slower)"));
+  m_radio_jit = new QRadioButton(tr("JIT Recompiler (recommended)"));
+  m_radio_jitil = new QRadioButton(tr("JITIL Recompiler (slow, experimental)"));
```
Comment (goldenercobra) [inline Source/Core/DolphinQt2/Settings/CPUPane.cpp]: ditto

File: Source/Core/DolphinQt2/Settings/CPUPane.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+  engine_group_layout->addWidget(m_radio_interpreter);
+  engine_group_layout->addWidget(m_radio_cached_interpreter);
+  engine_group_layout->addWidget(m_radio_jit);
+  engine_group_layout->addWidget(m_radio_jitil);
```
Comment (goldenercobra) [inline Source/Core/DolphinQt2/Settings/CPUPane.cpp]: ditto

File: Source/Core/DolphinQt2/Settings/CPUPane.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  case PowerPC::CPUCore::CORE_INTERPRETER:
+    m_radio_interpreter->setChecked(true);
+    break;
+  case PowerPC::CPUCore::CORE_CACHEDINTERPRETER:
+    m_radio_cached_interpreter->setChecked(true);
+    break;
+  case PowerPC::CPUCore::CORE_JIT64:
+    m_radio_jit->setChecked(true);
+    break;
+  case PowerPC::CPUCore::CORE_JITIL64:
```
Comment (goldenercobra) [inline Source/Core/DolphinQt2/Settings/CPUPane.cpp]: ditto

File: Source/Core/DolphinQt2/Settings/CPUPane.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  settings.bCPUThread = m_checkbox_multicore->isChecked();
+
+  int engine_value = 0;
+  if (m_radio_interpreter->isChecked())
+    engine_value = PowerPC::CPUCore::CORE_INTERPRETER;
+  else if (m_radio_cached_interpreter->isChecked())
+    engine_value = PowerPC::CPUCore::CORE_CACHEDINTERPRETER;
+  else if (m_radio_jit->isChecked())
+    engine_value = PowerPC::CPUCore::CORE_JIT64;
+  else if (m_radio_jitil->isChecked())
```
Comment (goldenercobra) [inline Source/Core/DolphinQt2/Settings/CPUPane.cpp]: ditto

**Comment by Ryanel** [comment]
Will work on this soon. Other then the few issues, should be RTM

CPU Clock override might end up here @myownfriend

**Comment by MayImilae** [comment]
The UI design within the WX UI has a lot of quirks and things we may want to iron out, but changing UI design requires a lot of bikeshedding and work from all parties. I don't think it's wise to be doing that while also moving to QT.

So, instead of a "CPU" tab, can you please just copy the General tab from the WX UI? From a design standpoint, it's already well ironed out and makes a bit of sense, plus everyone is used to it.

**Comment by JosJuice** [comment]
I agree with May. Keep large UI changes separate from the implementation of new parts of the UI when possible.

**Comment by Ryanel** [comment]
I will merge the panes, but i don't agree that the UI makes sense in it's current state either.
As developers, we know where things are placed and that generally things are in correct categories; To users, this is not the case. I've had several friends I've done netplay with, and they have been constantly confused about the layout. Why is netplay in tools? Where do I add my dumps? What the heck is "Movie". Why are there separate "things" for Graphics and Controllers? I've been asked all of these. Dolphin is certainly great, but it could be more user friendly.

**Comment by Ryanel** [comment]
@Helios747, you're right. I'll try and get it as close as possible.

**Comment by Helios747** [comment]
Note, moving from traditional tabs to the side panel stuff is fine. We're already committed to that so it's cool. But lets just copy over the "General" tab in this PR.

**Comment by Ryanel** [comment]
Qt: image
Wx:
image

**Comment by MayImilae** [comment]
Well, there are design differences, so time for a little bikeshedding!

Language shouldn't be here, simply because of this: what is that language being applied to? Dolphin? The emulated GameCube? The emulated Wii? Language for Dolphin is in the interface tab, language for the GC is in the GameCube tab, and language for the wii is in the Wii tab. That makes it very clear what is being used where, and allows for things like English Dolphin UI and Japanese Wii gameplay if one desires. This one is pretty important!

Speed Limit as a slider? Well... sliders are much less precise. There is a decent amount of precision here due to the massive width, but this window is enormous and I'm certainly wanting it to be narrowed considerably once more options are in place, and that will make the slider even less accurate. I think the existing dropdown is just more practical in this case.

**Comment by JosJuice** [comment]
To be fair, the language setting and speed slider were added in PR #5370, so those should probably not be blocking this PR.

**Comment by MayImilae** [comment]
Whoops, I wasn't on top of this. Still should be fixed though! This is definitely a design issue.

**Comment by Ryanel** [comment]
@MayImilae Please refer to PR #5370.
There were discussions on moving language to interface, but I raised usability concerns. 
To avoid confusion, I will simply change it to "UI Language" in another PR.

Additionally, the slider was requested as an alternative to the combobox. Looking at the old commit: 
image. I believe I was asked on IRC to make the change.

**However** you do raise a good point about the width. I will do this in another PR as well.

**Comment by MayImilae** [comment]
Well we were buying a car, so um, I was a busy!

I absolutely would have objected to this in pr5370 if I was aware of it. Regardless of being a little late, my point is definitely a valid one: Language should not be in general, because the items in general affect _everything_! JIT affects every game, speed limit affects every game, etc etc. Placing language here makes it quite unclear what the language selection is meant to apply to, and is a definite step back in design.

And whenever the window is narrowed, the slider will be much more limited compared to a dropdown.

Anyway, whether this is addressed in this PR or another one, I don't mind, but this is the general settings PR, so imo I think it should be addressed here.

**Comment by Ryanel** [comment]
@MayImilae
I went ahead with adjusting the maximum width of all stretched controls to 200.
Language has been moved to interface. It remains nonfunctional.
Speed has been reverted to a Combobox.

image
image

**Comment by myownfriend** [comment]
I'm not understanding why this PR changed to just copying the General tab. In the WX UI, we have CPU Emulation Engine in the  Advanced Settings area of the General tab, yet we also have a whole separate tab called Advanced that has a section called CPU Options.

In other words, Dolphin's own language infers that anything CPU related is an "Advanced" setting yet we relegate some to the Advanced tab and some to General. Since Custom RTC Options likely has next to no effect on most games and has no discernable warning associated with it, why not just swap it with CPU Emulation Engine, rename the Advanced tab to the CPU tab, and give it one all-encompassing warning saying something like "The settings in this tab will mess with speed and stability more than most settings in the UI."

I get that there is some worry that people will be confused when moving over to the new UI, but if things are in more logical locations then they were before, then it really shouldn't trip anyone up. The only way the having CPU Emulation Engine in the General tab would make sense to me is if CPU Clock Override was moved completely to the Game-Specific Settings.

As for the Interface tab above, I love the idea of taking some of the Interface options out of the Graphics settings and putting them in a centralized spot. I would also put Show FPS, Show NetPlay Messages, Show NetPlay Ping there, and maybe even Use Fullscreen there, too.

**Comment by JosJuice** [comment]
@myownfriend The reason for just copying DolphinWX is that we don't want to have a long-drawn discussion about the UX (like the one you seem to be trying to start right now). If we want any big changes like that, we should do it in a separate PR later (and preferably make the change in both DolphinQt2 and DolphinWX). Otherwise DolphinQt2 is going to take ages to catch up with DolphinWX.

**Comment by JosJuice** [comment]
@myownfriend 

> the Usage Statistics stuff shouldn't effect games at all and "Force as NTSC-J" shouldn't effect PAL games

Usage statistics are used while playing games (though it's invisible to the user unless they're sniffing the traffic), and Force as NTSC-J very much affects PAL games.

> Even if it were, that's not unique to the General tab. The settings in the Audio and Advanced tabs also effect every game.

Yes, and nobody has suggested putting the language option in the Audio or Advanced tabs.

**Comment by myownfriend** [comment]
@JosJuice  Fair enough. 

I get that no one suggested that the language setting be in those other tabs, I was just pointing out that we can't say that the criterion for an option being in the General tab is that it applies to all games since we can say that about most settings in other tabs as well. I'm aware that Usage Statistics do get used during games but they don't have any effect on emulation. If I understand the feature correctly, it's essentially a logging feature and effects games about as much as the settings in the Interface tab. 

As for Force NTSC-J, the description suggests it has no effect on PAL games, just NTSC-U and NTSC-J games. So when we are having a discussion about how to make the UI better, perhaps that should be explained.

**Comment by JosJuice** [comment]
> I get that no one suggested that the language setting be in those other tabs, I was just pointing out that we can't say that the criterion for an option being in the General tab is that it applies to all games since we can say that about most settings in other tabs as well.

May wasn't saying that all settings that affect all games are supposed to be in the General tab, she was saying that the settings that currently are in the General tab in DolphinWX affect all games.

> I'm aware that Usage Statistics do get used during games but they don't have any effect on emulation. If I understand the feature correctly, it's essentially a logging feature and effects games about as much as the settings in the Interface tab.

Sure, but it's not part of the interface, so putting it in the Interface tab doesn't make much sense. There isn't really any place it fits better in than General.

> As for Force NTSC-J, the description suggests it has no effect on PAL games, just NTSC-U and NTSC-J games. So when we are having a discussion about how to make the UI better, perhaps that should be explained.

It affects PAL and NTSC-U but not NTSC-J. This option should be entirely removed, to be honest. It's useful for pretty much nothing and is hard to understand.

**Comment by Ryanel** [comment]
This PR is pretty much complete. I recommend it be merged to prevent more bikeshedding.
I just kinda went along with what everyone wanted cause I didn't want any more bikeshedding.
HOWEVER, in the future, some of these ideas can be explored. Not here, though.

**Comment by MayImilae** [comment]
Much better! LGTM.

**Comment by Helios747** [comment]
Sorry, but this needs more UI fixing. This was on a fairly standard GNOME 3 install in Arch Linux

img

Checking though the code now

**Comment by Ryanel** [comment]
@Helios747 widened from 200 to 250. Should be good enough.

**Comment by lioncash** [comment]
@Ryanel is using explicit pixel widths a good solution for this sort of thing? Within Qt's [High DPI](https://doc.qt.io/qt-5/highdpi.html) documentation In the migration section for handling nicely with different screen DPI types, one of the recommendations are:


> Replace hard-coded sizes in layouts and drawing code by values calculated from font metrics or screen size.

**Comment by Ryanel** [comment]
@lioncash Didn't know this existed. I will use this (in another PR, converting any fixed widths to that)

File: Source/Core/DolphinQt2/Settings/GeneralPane.cpp [hunk 6]
```diff
@@ -86,6 +86,7 @@ void GeneralPane::CreateAnalytics()
 
   m_checkbox_enable_analytics = new QCheckBox(tr("Enable Usage Statistics Reporting"));
```
Comment (Helios747) [inline Source/Core/DolphinQt2/Settings/GeneralPane.cpp]: The 'g' at the end of reporting is still being cut off

**Comment by Helios747** [comment]
new comment because blegh @ github

The 'g' at the top, "usage statistics reporting" is still being cut off and needs to be fixed

**Comment by Helios747** [comment]
And move speed limit into basic settings

**Comment by Ryanel** [comment]
@Helios747 why?

**Comment by Helios747** [comment]
It's how it was in the old UI.

**Comment by Ryanel** [comment]
@Helios747 done.

**Pull request merged**